pkg/collector/corechecks/cluster/kubernetes_eventbundle.go

```diff
@@ -0,0 +1,87 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+// +build kubeapiserver
+
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/ericchiang/k8s/api/v1"
+)
+
+type kubernetesEventBundle struct {
+	objUid        string         // Unique object Identifier used as the Aggregation key
+	readableKey   string         // Formated key used in the Title in the events
+	component     string         // Used to identify the Kubernetes component which generated the event
+	events        []*v1.Event    // List of events in the bundle
+	timeStamp     float64        // Used for the new events in the bundle to specify when they first occurred
+	lastTimestamp float64        // Used for the modified events in the bundle to specify when they last occurred
+	countByAction map[string]int // Map of count per action to aggregate several events from the same ObjUid in one event
+}
+
+func newKubernetesEventBundler(objUid string, compName string) *kubernetesEventBundle {
+	return &kubernetesEventBundle{
+		objUid:        objUid,
+		component:     compName,
+		countByAction: make(map[string]int),
+	}
+}
+
+func (k *kubernetesEventBundle) addEvent(event *v1.Event) error {
+	if event == nil || event.InvolvedObject == nil {
+		return errors.New("could not retrieve some parent attributes of the event")
+	}
+	if *event.InvolvedObject.Uid != k.objUid {
+		return fmt.Errorf("mismatching Object UIDs: %s != %s", event.InvolvedObject.Uid, k.objUid)
+	}
+	k.events = append(k.events, event)
+	k.timeStamp = math.Max(k.timeStamp, float64(*event.Metadata.CreationTimestamp.Seconds))
+	k.lastTimestamp = math.Max(k.timeStamp, float64(*event.LastTimestamp.Seconds))
+
+	if event.Reason == nil || event.Message == nil || event.InvolvedObject.Name == nil || event.InvolvedObject.Kind == nil {
```

Comment:
Is it possible to move this on top ? Do we need to update k ?

Reply:
let's ignore the events that are not properly formatted and also ignore their contribution.
i.e. : if a new event is misformated, we don't want to have part of it's message in the countByAction for instance.